src/Microsoft.Diagnostics.Monitoring.EventPipe/DistributedTraces/TraceEventExtensions.cs

```diff
@@ -0,0 +1,223 @@
+                            }
+                            break;
+                        case "StatusDescription":
+                            statusDescription = value as string;
+                            break;
+                        case "Tags":
+                            string? tagsValue = value as string;
+                            if (!string.IsNullOrEmpty(tagsValue))
+                            {
+                                tagCount = ParseTags(tagsValue);
+                            }
+                            break;
+                        case "ActivitySourceVersion":
+                            sourceVersion = value as string;
+                            break;
+                    }
+                }
+
+                if (!s_Sources.TryGetValue(sourceName, out source))
+                {
+                    source = new(sourceName, sourceVersion);
+                    s_Sources[sourceName] = source;
+                }
+
+                payload.ActivityData = new ActivityData(
+                    source,
+                    activityName,
+                    displayName,
+                    kind,
+                    traceId,
+                    spanId,
+                    parentSpanId,
+                    traceFlags,
+                    traceState,
+                    startTimeUtc,
+                    startTimeUtc + TimeSpan.FromTicks(durationTicks),
+                    status,
+                    statusDescription);
+
+                payload.Tags = new(s_TagStorage, 0, tagCount);
+
+                return true;
+            }
+
+            payload = default;
+            return false;
+        }
+
+        private static int ParseTags(string tagsValue)
+        {
+            ref KeyValuePair<string, object?>[]? tags = ref s_TagStorage;
+            tags ??= new KeyValuePair<string, object?>[16];
+
+            int tagCount = 0;
+
+            for (int i = 0; i < tagsValue.Length; i++)
+            {
+                if (tagsValue[i++] != '[')
+                {
+                    break;
+                }
+
+                int commaPosition = tagsValue.IndexOf(',', i);
+                if (commaPosition < 0)
+                {
+                    break;
+                }
+
+                string key = tagsValue.Substring(i, commaPosition - i);
+
+                i = commaPosition + 2;
+
+                int endPosition = tagsValue.IndexOf(']', i);
+                if (endPosition < 0)
+                {
+                    break;
+                }
+
+                string value = tagsValue.Substring(i, endPosition - i);
+
+                i = endPosition + 1;
+
+                AddToArrayGrowingAsNeeded(ref tags, new(key, value), ref tagCount);
+            }
+
+            return tagCount;
+        }
+
+        private static void AddToArrayGrowingAsNeeded<T>(ref T[] destination, T item, ref int index)
+        {
+            if (index >= destination.Length)
```

Comment:
Consider making a separate helper method or extension method for growing the ref array. You have similar logic in the LogRecord PR.

Reply:
I pushed a tweak so the log code uses this helper.